keps/sig-storage/3751-volume-attributes-class/README.md

```diff
@@ -0,0 +1,1279 @@
+
+### Goals
+
+* Cloud-provider and storage vendor independent specification of basic volume attributes parameters
+* Volume attributes enforcement performed by the storage
+* Specifying volume attributes at volume creation
+* Changing volume attributes parameters on-line for existing volumes
+* Volume attributes modifiable non disruptively by workload developers
+* ResourceQuota enforcement by VolumeAttributesClass
+
+### Non-Goals
+
+* OS-level IO volume attributes, for example cgroup policy
+* Inter-pod volume attributes. This KEP is restricted to what attributes can be defined on a volume.
+* Scheduling based on total specific volume attributes limits of a node. For example, total throughput for many cloud provider network attached storage volumes is limited by instance network bandwidth and may prevent achieving the provisioned IO. Scheduling to mitigate these factors is out of scope but it should be considered as a future extension of this KEP.
+
+## Proposal
+
+### Kubernetes API
+We need to add a new resource object VolumeAttributesClass to Kubernetes API, also a new admission controller and vac protection controller. Please see more in [Design Details](#bookmark=id.wtvwymf8202g).
+
+The reason why we cannot use StorageClass.parameters is because StorageClass is immutable today, the design is keeping the naming opaque as mutable_parameters(mutable to the volume but immutable to the VolumeAttributesClass) because CSI was initially designed to support more than just Kubernetes, VolumeAttributesClass is similar to StorageClass, these are purely a Kubernetes-only concept and not a storage provider concept. 
+
+If there is conflict between StorageClass.parameters and VolumeAttributesClass.parameters, it is up to the driver to decide how to handle the conflict. The driver can either reject the request and return an error, or accept the parameters with its preferable way.
+
+
+```
+type VolumeAttributesClass struct {
+  metav1.TypeMeta
+  // +optional
+  metav1.ObjectMeta
+
+  // Spec defines the behavior of a VolumeAttributesClass.
+  // +optional
+  Spec VolumeAttributesClassSpec
+}
+
+type VolumeAttributesClassSpec struct {
+  // Name of the driver
+  DriverName string
+  // This field is OPTIONAL.This allows the CO to specify the
+  // volume attributes class parameters to apply. 
+  // These parameters are immutable 
+  Parameters map[string]string
+}
+
+type PersistentVolumeSpec struct {
+  ...
+  // Name of VolumeAttributesClass to which this persistent volume belongs. Empty value
+  // means that this volume does not belong to any VolumeAttributesClass.
+  // +optional
+  VolumeAttributesClassName string
+  ...
+}
+
+type PersistentVolumeClaimSpec struct {
+  ...
+  // Name of the VolumeAttributesClass required by the claim.
+  // More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes/#class-1
+  // +optional
+  VolumeAttributesClassName *string
```

Comment:
What about existing volumes though that don't have any VAC?

Reply:
Regardless we should verify the name? If it is an empty string, there is no VAC to retrieve?

Reply:
When StorageClasses were introduced (ages ago), we had to consider existing PVs/PVCs that were created when `pv.spec.storageClassName` field did not exists and only static provisioning was used. At that time, we've chosen to have `nil` in PVC and treat it as empty string during binding. Adding the default SC there could break binding of existing PVs/PVCs.

Today, we're adding code that treats `pvc.spec.storageClassName: nil` as "I still want the default SC if it's introduced later", because dynamic provisioning is now the major use case and static provisioning is not so popular (but important).

Just thinking out loud, I think we should *not* require `VolumeAttributesClassName` to be set nor a default VAC to be present - if an existing cluster gets upgraded to a version where VAC is GA, its users should not be forced to create the default VAC or change their PVC or StatefulSet yamls (helm charts, operators, ...)

I wonder if the default VAC should be retroactive from the start (i.e. existing PVCs with `vacName: nil` will get the default one *after* the default VAC is created - it could lead to lot of churn in the cluster with many PVCs.

Reply:
The situation of VAC is similar to that of StorageClass because we also need to consider existing PVs/PVCs that were created without `pv.spec.VolumeAttributesClassName` even if dynamic provisioning is more popular than static provisioning today. It will make the design simpler if we don't have to differentiate between nil and an empty string.
For existing volumes without VAC, VAC name is nil. 
Since we support retroactive default SC, I think we should also support that for VAC, but I suggest that we don't support VAC name to be "".  VAC name is optional so it can be nil. If a PVC without VAC name is already bound, we don't set it to the default VAC if it becomes available later. Users are not forced to create a default VAC after VAC becomes GA.

Reply:
I am adding this requirement of no empty string to the description.